presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java

```diff
@@ -345,38 +376,83 @@ private void print(int indent, String format, Object... args)
+    }
+
+    private void printStats(int indent, PlanNodeId planNodeId, boolean printInput, boolean printFiltered)
     {
         if (!stats.isPresent()) {
             return;
         }
+
         long totalMillis = stats.get().values().stream()
                 .mapToLong(node -> node.getWallTime().toMillis())
                 .sum();
-        PlanNodeStats stats = this.stats.get().get(planNodeId);
-        if (stats == null) {
-            output.append(indentString(indent))
-                    .append("Cost: unknown, Output: unknown \n");
+
+        PlanNodeStats nodeStats = stats.get().get(planNodeId);
+        if (nodeStats == null) {
+            output.append(indentString(indent));
+            output.append("Cost: ?");
+            if (printInput) {
+                output.append(", Input: ? lines (?B)");
+            }
+            output.append(", Output: ? lines (?B)");
+            if (printFiltered) {
+                output.append(", Filtered: ? %");
+            }
+            output.append('\n');
             return;
         }
-        double fraction = (stats.getWallTime().toMillis()) / (double) totalMillis;
+
+        double fraction = (nodeStats.getWallTime().toMillis()) / (double) totalMillis;
         String fractionString;
         if (isFinite(fraction)) {
-            fractionString = format("%.2f%%", 100.0 * fraction);
+            fractionString = format(Locale.US, "%.2f%%", 100.0 * fraction);
         }
         else {
-            fractionString = "unknown";
+            fractionString = "?";
+        }
+
+        output.append(indentString(indent));
+        output.append("Cost: " + fractionString);
+        if (printInput) {
+            output.append(format(", Input: %s (%s)",
+                    formatPositions(nodeStats.getInputPositions()),
+                    nodeStats.getInputDataSize().toString()));
+        }
+        output.append(format(", Output: %s (%s)",
+                formatPositions(nodeStats.getOutputPositions()),
+                nodeStats.getOutputDataSize().toString()));
+        if (printFiltered) {
+            double filtered = 100.0 * (nodeStats.getInputPositions() - nodeStats.getOutputPositions()) / nodeStats.getInputPositions();
+            String filteredString;
+            if (isFinite(filtered)) {
+                filteredString = format(Locale.US, "%.2f%%", filtered);
+            }
+            else {
+                filteredString = "?";
+            }
+
+            output.append(", Filtered: " + filteredString);
         }
+        output.append('\n');
+    }
 
-        String outputString;
-        if (stats.getOutputPositions().isPresent() && stats.getOutputDataSize().isPresent()) {
-            outputString = format("%s rows (%s)", stats.getOutputPositions().get(), stats.getOutputDataSize().get());
+    private static String formatPositions(long positions)
+    {
+        if (positions == 1) {
+            return "1 line";
```

Comment:
I believe the SQL standard refers to these as "rows", so this should be "row" rather than "line"